From 90fa04e61f7dc9a18e8e822c33a506272d4e51dc Mon Sep 17 00:00:00 2001
From: Wolfgang Frisch <wolfgang.frisch@suse.com>
Date: Mon, 29 Jun 2020 16:28:51 +0200
Subject: [PATCH] catch integer overflow in ext4_cache_extents

---
 fs/ext4/extents.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/fs/ext4/extents.c b/fs/ext4/extents.c
index 221f240eae60..052d50ec02a4 100644
--- a/fs/ext4/extents.c
+++ b/fs/ext4/extents.c
@@ -471,6 +471,9 @@ static void ext4_cache_extents(struct inode *inode,
 		ext4_lblk_t lblk = le32_to_cpu(ex->ee_block);
 		int len = ext4_ext_get_actual_len(ex);
 
+		if (lblk + len < lblk)
+			return;
+
 		if (prev && (prev != lblk))
 			ext4_es_cache_extent(inode, prev, lblk - prev, ~0,
 					     EXTENT_STATUS_HOLE);
-- 
2.27.0

